### ISSUE TITLE
MOS calculation bug

### ISSUE THREAD

**Issue by mtryfoss** [sipcapture/homer#258]
Hi!

I've been investigating a number of calls with really bad MOS scores, which obviously isn't that bad in real life. After studying the calculateJitterMos function, I'm pretty sure there's an error here related to packet loss.

I use Asterisk's hep module and this reports cumulative packet loss to Sipcapture. This is fed into the calculation function. Could this be correct? If there are a small amount of packet loss (~3%) now and then this will increase over time and never let the MOS value rise again. When checking each report, the score fell to only 1 when cumulative number of packet loss reached 20 and never rised again.

Tried modifying the code so I'm using packet loss since last report for a given SSRC, and this seems more correct. Now the MOS score raised from 1.5 to 4.2.

Can anyone confirm this or give me any suggestions? :)

**Comment by lmangani** [comment]
@mtryfoss thanks for the report and details. Could you share the code changes you've attempted in a PR we can all improve together?

**Comment by mtryfoss** [comment]
Of course I can, but this was more or less a blind shot. I'm no expert in this field, and hoped somebody could confirm or not whenever this is the correct assumptions before I do submit code.

**Comment by lmangani** [comment]
@mtryfoss don't be shy, we're all Humans here learning from each other - we will refine whatever you started as a community in the PR. If you prefer, send the current changes privately to support@sipcapture.org for discussion.

**Comment by mtryfoss** [comment]
I also remembered that RTPEngine got a MOS calculation algorithm. It seems like they're using the "fraction_lost" instead of cumulative:

Code:
packetloss = (unsigned int) rr->fraction_lost * 100 / 256

According to an example trace, this will give even better MOS score than using the delta value of cumulative.

**Comment by lmangani** [comment]
I'm tempted to add this as an option so we can freely play with those approaches (breaking-or-else) and let users choose what to use until we have a winner? Comments?

**Comment by mtryfoss** [comment]
The original approach seems to be wrong anyway. 

I've extracted the code and run it through several thousands of calls using a custom script. I got a lot of long-lasting calls with score as low as 1 for some periods. You're probably not able to do any conversation if that was real. After my adjustments, this has improved a lot and seems more realistic.

The question is, what's the best way to correct it?

**Comment by mtryfoss** [comment]
RTPEngine's approach:

`$tmpMos = round(calculateJitterMos($block["dlsr"] < 1000 ? $block["dlsr"] : 0 ,$block["ia_jitter"],($block["fraction_lost"] * 100 / 256)),2);`

**Comment by negbie** [comment]
If you use "fraction_lost" keep in mind  that a receiver cannot tell whether any packets were lost after the last one received, and that there will be no reception report block issued for a source if all packets from that source sent during the last reporting interval have been lost.

So no matter what you will use here it will never be 100% accuarate. IMHO I would use a delta between cumulative packet loss. But nevertheless this function will alwas be fishy because of:
this argument ($block["dlsr"] < 1000 ? $block["dlsr"] : 0 )
and inside the function ($rtt == 0) $rtt = 10;

RTT will quite lineraly degrade MOS at 200 ms for 1 MOS point but we don't have good RTT values here.

**Comment by lmangani** [comment]
I agree with both honestly, this is why I was suggesting to simply allow more approaches/formulas to co-exist and let the user choose the one producing the more realistic averaging for their network. Happy to default to this new one, clearly if we have a quorum. Thanks @mtryfoss for taking the time to test this in depth, its a great contribution to the overall picture - although as per @negbie comments which are 100% spot-on, we can really only guess-timate things without a proper RTT measurement per interval.

**Comment by negbie** [comment]
@lmangani: yip agree ;) Since it's broken anyway we can use @mtryfoss suggestion.
@mtryfoss: if you want to do some kind of alerting or rely on rtcp values I would recommend you the homer metric branch. For example you could graph the jitter, cum_packet_lost, packet_lost like this:
![rtcp_metrics](https://user-images.githubusercontent.com/20154956/32327402-af8f817e-bfd6-11e7-811e-d758e043c13c.jpg)

In Kapacitor you can aggregate what ever you want. For example look for jitter > 50ms and high packet_lost and send you those call-id's with srcIP+dstIP. If you go one step further and combine this with elasticsearch or graylog you won't ever need the QoS Report Tab.

**Comment by mtryfoss** [comment]
Thanks for the suggestions! We're still using a very old version of homer, but the plan is to migrate to a newer version soon. 

I've only installed a new capture node as a proof of concept to be able to grab RTCP info from some Asterisk nodes, and analyze the data by some custom developed scripts. Used the calculation function from Homer, and then I discovered this bug.

**Comment by lmangani** [comment]
@mtryfoss you plan sending a PR for this? Thanks!
